File: packages/python/goldencheck/goldencheck/denial/validate.py
"""Ground-truth g1 validation for candidate denial constraints.

Discovery (evidence + :func:`discover`) works on a *sample* and emits candidate
DCs as predicate conjunctions. This module re-measures each candidate's g1 on the
real Polars frame so a sample artefact never ships as a finding.

A DC ``¬(p1 ∧ … ∧ pm)`` is VIOLATED by an element when ALL of ``p1..pm`` hold:

* **Single-tuple** (every predicate references one row -- kind ``const``/``single``):
  validated EXACTLY over all n rows, O(n). ``g1 = |violating rows| / n`` and the
  exact violating row indices are returned.
* **Cross-tuple** (at least one ``cross`` predicate): full-table validation is
  O(n^2), so we validate on a bounded, seeded row sample and return an ESTIMATED
  g1 over the sample's ordered pairs plus a few representative violating pairs.
"""
from __future__ import annotations

import random

import polars as pl

from goldencheck.denial.constants import VALIDATION_SAMPLE
from goldencheck.denial.models import Predicate
from goldencheck.denial.predicates import encode_columns, predicate_holds

__all__ = ["is_single_tuple", "validate_single_tuple", "validate_cross_tuple"]


def is_single_tuple(preds: list[Predicate]) -> bool:
    """True iff no predicate has kind ``"cross"`` (the DC scopes to one tuple)."""
    return all(p.kind != "cross" for p in preds)


def validate_single_tuple(
    preds: list[Predicate], df: pl.DataFrame
) -> tuple[float, list[int]]:
    """Exact O(n) validation. Returns ``(g1, violating_row_indices)``.

    A row violates when ALL ``preds`` hold on it (every pred is const/single, so
    ``row_b=None``). ``g1 = len(violating) / df.height``.
    """
    n = df.height
    if n == 0:
        return 0.0, []
    enc = encode_columns(df)
    violating = [
        r for r in range(n) if all(predicate_holds(p, enc, r, None) for p in preds)
    ]
    return len(violating) / n, violating


def validate_cross_tuple(
    preds: list[Predicate],
    df: pl.DataFrame,
    *,
    sample: int = VALIDATION_SAMPLE,
    seed: int = 0,
    max_pairs: int = 5,
) -> tuple[float, list[tuple[int, int]]]:
    """Estimated g1 over a bounded, seeded row sample. Returns ``(g1_est, pairs)``.

    Takes ``min(df.height, sample)`` rows (seeded), evaluates every ordered pair
    ``(α, β)`` with ``α != β``. A pair violates when ALL ``preds`` hold: const/single
    preds evaluated on ``α`` (row_b=None), cross preds on ``(α, β)``.
    ``g1_est = violations / (m*(m-1))`` where ``m`` is the sample size. Returns up
    to ``max_pairs`` representative violating ``(α, β)`` index pairs into the SAMPLE.
    """
    n = df.height
    m = min(n, sample)
    if m < 2:
        return 0.0, []

    if n <= sample:
        rows = list(range(n))
    else:
        rng = random.Random(seed)
        rows = rng.sample(range(n), m)

    sub = df[rows]
    enc = encode_columns(sub)

    single = [p for p in preds if p.kind != "cross"]
    cross = [p for p in preds if p.kind == "cross"]

    violations = 0
    examples: list[tuple[int, int]] = []
    for a in range(m):
        if not all(predicate_holds(p, enc, a, None) for p in single):
            continue  # single-tuple part fails on α -> no pair (α,·) can violate
        for b in range(m):
            if a == b:
                continue
            if all(predicate_holds(p, enc, a, b) for p in cross):
                violations += 1
                if len(examples) < max_pairs:
                    examples.append((a, b))

    g1_est = violations / (m * (m - 1))
    return g1_est, examples

File: packages/typescript/goldencheck/src/core/denial/validate.ts
/**
 * Ground-truth g1 validation for candidate denial constraints. Parity port of
 * `packages/python/goldencheck/goldencheck/denial/validate.py`.
 *
 * Discovery works on a sample; this re-measures each candidate's g1 on the real
 * frame so a sample artefact never ships as a finding.
 */

import { TabularData } from "../data.js";
import { VALIDATION_SAMPLE } from "./constants.js";
import type { Predicate } from "./models.js";
import { encodeColumns, predicateHolds } from "./predicates.js";

/** True iff no predicate has kind "cross" (the DC scopes to one tuple). */
export function isSingleTuple(preds: readonly Predicate[]): boolean {
  return preds.every((p) => p.kind !== "cross");
}

/** Exact O(n) validation. Returns `[g1, violatingRowIndices]`. */
export function validateSingleTuple(preds: readonly Predicate[], df: TabularData): [number, number[]] {
  const n = df.rowCount;
  if (n === 0) return [0.0, []];
  const enc = encodeColumns(df);
  const violating: number[] = [];
  for (let r = 0; r < n; r++) {
    if (preds.every((p) => predicateHolds(p, enc, r, null))) violating.push(r);
  }
  return [violating.length / n, violating];
}

/** Build a sub-frame from selected row indices. */
function subFrame(df: TabularData, rows: readonly number[]): TabularData {
  return new TabularData(rows.map((r) => df.rows[r]!));
}

/**
 * Estimated g1 over a bounded row sample. Returns `[g1Est, pairs]`.
 *
 * DIVERGENCE: when `n > sample` Python draws a seeded `random.Random(seed).sample`
 * whose Mersenne-Twister sequence TS cannot reproduce; TS falls back to the first
 * `sample` rows. This only fires above `VALIDATION_SAMPLE` (20000) rows — the
 * parity fixture stays well under it, so `rows = range(n)` and both are identical.
 */
export function validateCrossTuple(
  preds: readonly Predicate[],
  df: TabularData,
  opts: { sample?: number; maxPairs?: number } = {},
): [number, Array<[number, number]>] {
  const sample = opts.sample ?? VALIDATION_SAMPLE;
  const maxPairs = opts.maxPairs ?? 5;
  const n = df.rowCount;
  const m = Math.min(n, sample);
  if (m < 2) return [0.0, []];

  const rows = n <= sample ? Array.from({ length: n }, (_, i) => i) : Array.from({ length: m }, (_, i) => i);
  const sub = subFrame(df, rows);
  const enc = encodeColumns(sub);

  const single = preds.filter((p) => p.kind !== "cross");
  const cross = preds.filter((p) => p.kind === "cross");

  let violations = 0;
  const examples: Array<[number, number]> = [];
  for (let a = 0; a < m; a++) {
    if (!single.every((p) => predicateHolds(p, enc, a, null))) continue;
    for (let b = 0; b < m; b++) {
      if (a === b) continue;
      if (cross.every((p) => predicateHolds(p, enc, a, b))) {
        violations++;
        if (examples.length < maxPairs) examples.push([a, b]);
      }
    }
  }

  const g1Est = violations / (m * (m - 1));
  return [g1Est, examples];
}
